src/codegen/enums.rs

```diff
@@ -68,6 +73,56 @@ pub fn generate(env: &Env, root_path: &Path, mod_rs: &mut Vec<String>) {
     });
 }
 
+fn generate_static_to_str(
+    env: &Env,
+    w: &mut dyn Write,
+    enum_: &Enumeration,
+    function: &crate::analysis::functions::Info,
+) -> Result<()> {
+    writeln!(w)?;
+    version_condition(w, env, function.version, false, 1)?;
+
+    // TODO: use the normal to_string name, and pass _by ref_!
+    // The trait_impls Display impl calls this... But it'll recursively
+    // call ToString::to_string -> Display::fmt if this function were by reference.
+    writeln!(w, "\tfn to_string(&self) -> &'static str {{")?;
+
+    let unk_match = if enum_.members.iter().any(|m| m.name == "unknown") {
+        "Self::Unknown | "
+    } else {
+        ""
+    };
+
+    writeln!(
+        w,
+        "\
+\t\tmatch self {{
+\t\t\t{}Self::__Unknown(_) => return \"UNKNOWN\",
+\t\t\t_ => {{}},
+\t\t}}",
+        unk_match
+    )?;
+
+    // TODO: Can we somehow reuse function_body_chunk::Builder here?
+
+    writeln!(
+        w,
+        "\
+\t\tunsafe {{
+\t\t\tlet ptr = {}::{}(self.to_glib());
+\t\t\tif ptr.is_null() {{
+\t\t\t\treturn \"INVALID\";
```

Comment:
Just like `to_string()`, this should probably panic if it's `NULL` unless the return value is marked as `nullable` in the introspection data. If it is marked as `nullable`, then we should probably **not** generate this at all and print a warning. It will require case by case specific manual implementations.

Reply:
> Just like to_string(), this should probably panic if it's NULL

Like the implicit panic through an assertion in `from_glib_none` -> `CStr::from_ptr`?

> unless the return value is marked as nullable in the introspection data

Hmm:

https://github.com/gtk-rs/gir/blob/062b3d649ccc09c17564d6bc429dc682829d71f0/src/analysis/special_functions.rs#L45-L48

As mentioned before this is a hacky approach, perhaps I should pass down the surrounding type to `special_functions::extract` and make a `StaticToString` (activated on flags/enum) to generate `impl Display` that leverages `.to_str()` instead. Still like `.to_str()` itself to show up in the original `impl` block though.

> It will require case by case specific manual implementations.

Well, it was the goal to generate this since all implementations thus far are identical. Should I abandon this effort instead?

Reply:
> Like the implicit panic through an assertion in from_glib_none -> CStr::from_ptr?

For example

> Well, it was the goal to generate this since all implementations thus far are identical. Should I abandon this effort instead?

You mean the example in GStreamer that handles `Unknown` as `"UNKNOWN"`? That's not a general GLib pattern so I don't think we should automatically do that.

Reply:
> You mean the example in GStreamer that handles Unknown as "UNKNOWN"? That's not a general GLib pattern so I don't think we should automatically do that.

The entire `to_str` function: name, dealing with invalid values upfront that might cause the C function to return `NULL`, and returning a static lifetime borrow.

If this is non-standard I can revert most of the code. That means the original `to_string(&self) -> GString` functions are going to pop back up. I don't really know what direction we want to go with these functions.

Reply:
`to_str()` for cases that are not nullable and transfer none is standard. `to_string()` for not nullable and transfer full is also standard. That should cover 99% of the cases and can be autogenerated.

nullable needs special handling per case, but we can create a `to_str()` / `to_string()` function that returns an `Option` I guess but that also means that it can't be used for automatically implementing `Display` / `ToString` (which is fine). Autogenerating that should be good, and if something better can be done it can be ignored in `Gir.toml` and a manual implementation can be done.

Also non-standard is the weird GStreamer case where one of the enum variants always returns `NULL` and that definitely needs manual handling.

Reply:
All functions handled by this bit of code have `transfer none` return type and are nullable. They all reside in gstreamer, gtk doesn't seem to have any.  Bitfields however have only one such function (in `GLSLProfile`), the other `to_string`s are all transfer full, not nullable and get the default `to_string -> GString` implementation (`GLAPI` and `GLPlatform`).

> nullable needs special handling per case, [...]

Given the above distribution of functions in gstreamer, what do you want to do exactly? We only have non-nullable variants in bitflags, and you seem to suggest to turn all the others into `Option<&'static str>` without `Display`?

Upfront warning: In the current push all enum/bitfield instance parameters are passed by value. This means `fn to_string` (for the transfer full case) is not found by the autogenerated `f.write_str(&self.to_string())` in `Display` resulting in aforementioned infinite recursion (caught by clippy at least). Perhaps we should rename _all_ `to_string` special functions to `to_str` and be done with this possibility forever?

> Also non-standard is the weird GStreamer case where one of the enum variants always returns NULL and that definitely needs manual handling.

You mean the `return_if_false` case that gets compiled out on release? In debug calling `to_string(unknown)` returns `NULL` while in release it returns `"unknown"`.

Reply:
> Given the above distribution of functions in gstreamer, what do you want to do exactly? We only have non-nullable variants in bitflags, and you seem to suggest to turn all the others into Option<&'static str> without Display?

All others are `nullable`, really? I would've assumed the `nullable` case to be very rare... but yes, for `nullable` we can't really autogenerate a sensible `Display` impl as it's not clear what the `NULL` case should represent.

> Perhaps we should rename all to_string special functions to to_str and be done with this possibility forever?

Or use the full function call syntax for calling it. `Type::fmt(self, f)` vs `Type::fmt(&self, f)`. That seems safer anyway and we should probably also do that in more places, also for `<Type as Trait>::function()` things.

> You mean the return_if_false case that gets compiled out on release? In debug calling to_string(unknown) returns NULL while in release it returns "unknown".

`g_return_val_if_fail()` cases are to be transformed (manually...) into panics on the Rust side and don't have to be considered for the signature, for example. We also assume that they're not compiled out for bindings but always trigger (compiling them out is not the default for releases and you have to explicitly do that, and then you'll have a lot of fun if you're not careful).

Reply:
> All others are `nullable`, really? I would've assumed the `nullable` case to be very rare... but yes, for `nullable` we can't really autogenerate a sensible `Display` impl as it's not clear what the `NULL` case should represent.

Yup, every `transfer none` on flags and enums is nullable currently. I guess we can assume either a panic or `"NONE"`/`"UNKNOWN"` here as is done currently in the code?

> Or use the full function call syntax for calling it. `Type::fmt(self, f)` vs `Type::fmt(&self, f)`. That seems safer anyway and we should probably also do that in more places, also for `<Type as Trait>::function()` things.

I guess that's a bit hard to do inside the `fn fmt` `impl Display for ...` itself, isn't it?

> `g_return_val_if_fail()` cases are to be transformed (manually...) into panics on the Rust side and don't have to be considered for the signature, for example. We also assume that they're not compiled out for bindings but always trigger (compiling them out is not the default for releases and you have to explicitly do that, and then you'll have a lot of fun if you're not careful).

That looks to be the only reason for `NULL` together with out-of-bounds values. Basically the code backup I have now to `return "UNKNOWN"` on `matches!(self, Unknown, __Unknown(_))` and on a return value of `NULL` should both panic?

Reply:
> Yup, every transfer none on flags and enums is nullable currently. I guess we can assume either a panic or "NONE"/"UNKNOWN" here as is done currently in the code?

Is it really `nullable` or only because the annotation is ignored? See my other comment.

Assertions/`g_return_val_if_fail()` are not considered `nullable`, and the bound checks also not as they're impossible in proper languages unlike C. You can ignore these cases and if a function is marked `nullable` just because of these cases then that should be fixed.